src/CGIHandler.cpp

```diff
@@ -0,0 +1,371 @@
+	if (extensionStart == std::string::npos)
+		return "";
+	std::string tmp = path.substr(extensionStart);
+	size_t pathInfoStart = tmp.find('/');
+	if (pathInfoStart == std::string::npos)
+		return "";
+	return tmp.substr(pathInfoStart);
+}
+
+/**
+ * @brief Extracts the script path from the given path based on the CGI extension.
+ *
+ * This function searches for the CGI extension within the provided path and
+ * returns the substring from the beginning of the path up to and including the
+ * CGI extension. If the CGI extension is not found, an empty string is returned.
+ *
+ * @param path The full path from which to extract the script path.
+ * @return std::string The extracted script path or an empty string if the CGI extension is not found.
+ */
+std::string CGIHandler::extractScriptPath(const std::string& path)
+{
+	size_t extensionStart = path.find(m_cgiExt);
+	if (extensionStart == std::string::npos)
+		return "";
+	return path.substr(0, extensionStart + m_cgiExt.size());
+}
+
+/**
+ * @brief Registers signal handlers for child processes.
+ *
+ * This function attempts to register signal handlers for specific signals
+ * that are relevant to child processes. Currently, it registers the handler
+ * for SIGQUIT to be ignored.
+ *
+ * @return true if the signal handlers were successfully registered, false otherwise.
+ */
+bool registerChildSignals()
+{
+	// NOLINTNEXTLINE(cppcoreguidelines-pro-type-cstyle-cast, performance-no-int-to-ptr)
+	if (std::signal(SIGQUIT, SIG_IGN) == SIG_ERR) {
+		LOG_ERROR << "failed to register " << signalNumToName(SIGQUIT) << ": " << std::strerror(errno);
+		return (false);
+	}
+	return (true);
+}
+
+/**
+ * @brief Closes the input and output pipes used by the CGI handler.
+ *
+ * This function closes both ends of the input and output pipes
+ * associated with the CGI handler. It ensures that all file
+ * descriptors related to the pipes are properly closed to prevent
+ * resource leaks.
+ */
+void CGIHandler::closePipes()
+{
+	webutils::closePipeEnd(m_pipeIn[0]);
+	webutils::closePipeEnd(m_pipeIn[1]);
+	webutils::closePipeEnd(m_pipeOut[0]);
+	webutils::closePipeEnd(m_pipeOut[1]);
+}
+
+/**
+ * @brief Closes all file descriptors associated with the epoll instance and active connections.
+ *
+ * This function ensures that all pipes, the epoll file descriptor, and any file descriptors
+ * associated with active connections in the given map are properly closed.
+ *
+ * @param epollFd The file descriptor for the epoll instance to be closed.
+ * @param connections A map containing connection file descriptors as keys and
+ *        corresponding `Connection` objects as values. All connection file descriptors will be closed.
+ *
+ */
+void CGIHandler::closeAllFds(int epollFd, std::map<int, Connection> connections)
+{
+	closePipes();
+	close(epollFd);
+	for (std::map<int, Connection>::iterator iter = connections.begin(); iter != connections.end(); ++iter)
+		close(iter->first);
```

Comment:
We still have a problem: _other_ pipes could be open.

We can't - what I initially thought - pass CLO_EXEC in pipe(), we would need pipe2() for that.

We could make a destructor for Connection() which does a slimmed down version of cleanConnection(): just close the pipe fds. 
Then in the connection loop close iter->first and remove element (would need other style of loop for that since iterator would be invalidated)

Or does the connection not track all open CGI pipes?

Reply:
Each Connection only contains the pipes belonging to itself, not those of the other Connections.
I am passing now the map of pipes/Connection* to the execute() function of the CGIHandler which contains all open pipes. With the help of this map, all open pipes are being closed now. I think this should be it